server/src/business-layer/services/StripeService.ts

```diff
@@ -22,6 +36,87 @@ export default class StripeService {
     return result.data
   }
 
+  /**
+   *
+   * @param email ideally extracted from JWT
+   * @param uid ideally extracted from JWT
+   */
+  public async createNewUser(email: string, uid: string) {
+    return await stripe.customers.create({
+      email,
+      metadata: { [USER_ID_KEY]: uid }
+    })
+  }
+
+  /**
+   * Checks for processing payments from a user. Used to avoid
+   * users from double paying after completing a checkout session for membership payments.
+   *
+   * @warning This assumes that users won't be paying for membership right after booking
+   *
+   * @param customerId `stripe_id` from the firebase document
+   */
+  public async hasProcessingPaymentIntent(customerId: string) {
+    const { data } = await stripe.paymentIntents.list({
+      customer: customerId
+    })
+    const hasProcessingPaymentIntent = !!data.find(
+      (intent) => intent.status === "processing"
+    )
+
+    return hasProcessingPaymentIntent
+  }
+
+  /**
+   *
+   * Checks for processing payments from a user. Used to avoid
+   * users from double paying after completing a checkout session for membership payments.
+   *
+   * @param customerId `stripe_id` from the firebase document
+   * @param createdMinutesAgo how long ago to check for checkout sessions
+   * @returns true if user has completed checkout session from a default `2` minutes ago
+   */
+  public async hasRecentlyCompletedCheckoutSession(
```

Comment:
I understand the intent of this function (we are attempting to make sure that Stripe posts to our webhook before a user attempts to pay for membership again), but I'm unsure if this is necessary.

The docs for [webhooks on `PaymentIntent`](https://docs.stripe.com/payments/payment-intents/verifying-status#webhooks) makes me think that Stripe provides some level of guarantee that the API won't return a `complete` session until _after_ the webhook has been sent.

We probably need to explore the guarantees of the webhooks from Stripe (e.g., what happens if our backend goes down and a user completes a session? How will Stripe re-deliver the webhook event) to investigate if this function is needed.

Reply:
Stripe seems to retry the webhook for up to 3 days: https://docs.stripe.com/webhooks#retries

Reply:
I was thinking its more to account for how we have to also refresh user tokens on the front end (to check if their account has been updated) so they might have a window to accidentally make another call - but it's probably true their first instinct will not to be pay again.

Reply:
If we remove this the claims check in this endpoint would be the only safety net against this

Reply:
> I was thinking its more to account for how we have to also refresh user tokens on the front end (to check if their account has been updated) so they might have a window to accidentally make another call - but it's probably true their (user's) first instinct will not to be pay again.

With this current implementation, I don't think this will be allowed by the server if the webhook is processed at the correct time (as you say, the claims check in the endpoint).

Our server uptime should be pretty good and probably would give sufficient confidence to not need to do this check (and instead rely on Stripe delivering the webhook event between server crashes) -- i.e., the server will correctly reject the request.

I'd say we leave this for now, as it's not going to do any harm (it's only 1 extra API call to Stripe).

Reply:
one thing to note is WDCC have deployed our app on a serverless platform, idk how that'll affect it